File: packages/dart/lib/src/bbox/bounding_box.dart
class BoundingBox {
  final double minX;
  final double minY;
  final double maxX;
  final double maxY;

  const BoundingBox({
    required this.minX,
    required this.minY,
    required this.maxX,
    required this.maxY,
  });

  Map<String, double> toMap() => {
        'minX': minX,
        'minY': minY,
        'maxX': maxX,
        'maxY': maxY,
      };
}

BoundingBox boundingBox(Map<String, dynamic> geojson) {
  var minX = double.infinity;
  var minY = double.infinity;
  var maxX = double.negativeInfinity;
  var maxY = double.negativeInfinity;

  void visitCoord(List<dynamic> coord) {
    final x = (coord[0] as num).toDouble();
    final y = (coord[1] as num).toDouble();
    if (x < minX) minX = x;
    if (y < minY) minY = y;
    if (x > maxX) maxX = x;
    if (y > maxY) maxY = y;
  }

  void scan(dynamic node) {
    if (node is List) {
      if (node.isNotEmpty && node[0] is num) {
        visitCoord(node);
      } else {
        for (final child in node) {
          scan(child);
        }
      }
    }
  }

  void walk(Map<String, dynamic> obj) {
    final type = obj['type'];
    if (type == 'FeatureCollection') {
      for (final feature in obj['features'] as List<dynamic>) {
        walk(feature as Map<String, dynamic>);
      }
    } else if (type == 'Feature') {
      final geometry = obj['geometry'];
      if (geometry != null) walk(geometry as Map<String, dynamic>);
    } else if (type == 'GeometryCollection') {
      for (final geometry in obj['geometries'] as List<dynamic>) {
        walk(geometry as Map<String, dynamic>);
      }
    } else {
      scan(obj['coordinates']);
    }
  }

  walk(geojson);

  return BoundingBox(minX: minX, minY: minY, maxX: maxX, maxY: maxY);
}

File: packages/typescript/src/bbox/boundingBox.ts
import type { BoundingBox } from "../types";

type GeoJson = any;

function walk(coords: GeoJson, acc: BoundingBox): void {
  if (Array.isArray(coords)) {
    if (
      coords.length >= 2 &&
      typeof coords[0] === "number" &&
      typeof coords[1] === "number"
    ) {
      const [x, y] = coords as [number, number];
      if (x < acc.minX) acc.minX = x;
      if (x > acc.maxX) acc.maxX = x;
      if (y < acc.minY) acc.minY = y;
      if (y > acc.maxY) acc.maxY = y;
      return;
    }
    for (const child of coords) {
      walk(child, acc);
    }
  }
}

function collect(node: GeoJson, acc: BoundingBox): void {
  if (!node || typeof node !== "object") {
    return;
  }
  switch (node.type) {
    case "FeatureCollection":
      for (const feature of node.features ?? []) {
        collect(feature, acc);
      }
      return;
    case "Feature":
      collect(node.geometry, acc);
      return;
    case "GeometryCollection":
      for (const geometry of node.geometries ?? []) {
        collect(geometry, acc);
      }
      return;
    default:
      if (node.coordinates !== undefined) {
        walk(node.coordinates, acc);
      }
  }
}

export function boundingBox(geojson: GeoJson): BoundingBox {
  const acc: BoundingBox = {
    minX: Infinity,
    minY: Infinity,
    maxX: -Infinity,
    maxY: -Infinity,
  };
  collect(geojson, acc);
  return acc;
}
